### DISCUSSION TITLE
How Geo Location simulation works in Appium?

### DISCUSSION THREAD

**Discussion by itkhanz** [appium/appium#20531] [Q&A] [answered]
I am trying to understand the internals of how setting the geo location works with Appium?

For Android, here are the Appium Server Logs when I tried to simulate GPS location via Appium Inspector on emulator. 

```
[HTTP] --> POST /session/f3caa791-9a3c-423e-ab94-0c1689db7ef0/location {"location":{"latitude":48.177027,"longitude":11.591614}}
[AndroidUiautomator2Driver@65c5] Calling AppiumDriver.setGeoLocation() with args: [{"latitude":48.177027,"longitude":11.591614},"f3caa791-9a3c-423e-ab94-0c1689db7ef0"]
[ADB] Overriding /Users/xxxx/.emulator_console_auth_token with an empty string to avoid telnet authentication for emulator commands
[ADB] Running '/Users/xxxx/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 emu geo fix 11.59162 48.17703'
[ADB] Running '/Users/xxxx/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 emu geo fix 11,59162 48,17703'
[ADB] Running '/Users/xxxx/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell dumpsys activity services io.appium.settings'
[ADB] Running '/Users/xxxx/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am broadcast -n io.appium.settings/.receivers.LocationInfoReceiver -a io.appium.settings.location'
[SettingsApp] Got geo coordinates: {"latitude":"37.4219983","longitude":"-122.0840000","altitude":"5.0000000"}
[AndroidUiautomator2Driver@65c5] Responding to client with driver.setGeoLocation() result: {"latitude":37.4219983,"longitude":-122.084,"altitude":5}
[HTTP] <-- POST /session/f3caa791-9a3c-423e-ab94-0c1689db7ef0/location 200 137 ms - 67
```

Similar logs are also for real device.
```
[HTTP] --> POST /session/75387387-382b-4c82-9f65-b18a677b30ed/location {"location":{"latitude":48.177027,"longitude":11.591614}}
[AndroidUiautomator2Driver@2bb3] Calling AppiumDriver.setGeoLocation() with args: [{"latitude":48.177027,"longitude":11.591614},"75387387-382b-4c82-9f65-b18a677b30ed"]
[ADB] Running '/Users/ibtisam.khan/Library/Android/sdk/platform-tools/adb -P 5037 -s ABC123456 shell am start-foreground-service -e longitude 11.59162 -e latitude 48.17703 io.appium.settings/.LocationService'
[ADB] Running '/Users/ibtisam.khan/Library/Android/sdk/platform-tools/adb -P 5037 -s ABC123456 shell dumpsys activity services io.appium.settings'
[ADB] Running '/Users/ibtisam.khan/Library/Android/sdk/platform-tools/adb -P 5037 -s ABC123456 shell am broadcast -n io.appium.settings/.receivers.LocationInfoReceiver -a io.appium.settings.location'
[SettingsApp] Got geo coordinates: {"latitude":"48.1766060","longitude":"11.5928156","altitude":"576.3999634"}
[AndroidUiautomator2Driver@2bb3] Responding to client with driver.setGeoLocation() result: {"latitude":48.176606,"longitude":11.5928156,"altitude":576.3999634}
[HTTP] <-- POST /session/75387387-382b-4c82-9f65-b18a677b30ed/location 200 338 ms - 78
```

After the location is applied, I get this notification from Android System that `Mock location is on. You're using a mock location from Appium Settings.`

Could you please help me understand how Appium Settings app and these ADB commands are working to simulate the geo location? Do they change the location settings of device under test or only work to simulate location settings for actual app under test? What is the role of Appium Settings app here?

I could not find any documentaion on this except some unit tests and end points documented here:
https://github.com/appium/java-client/blob/34a8d5f154f7337db400e2566e643c1ea4052ae3/src/e2eAndroidTest/java/io/appium/java_client/android/AndroidDriverTest.java#L194

https://github.com/appium/appium/blob/f5f0aa5068ac6a02c422273e9e857152fea1675a/packages/base-driver/docs/mjsonwp/protocol-methods.md?plain=1#L79C1-L80C109

https://appium.github.io/appium.io/docs/en/commands/session/geolocation/set-geolocation/

**Comment by mykola-mokhnach** [comment, marked as answer]
Both emulators and real devices use different approach to emulate geolocation.
Emulators utilize https://stackoverflow.com/questions/2279647/how-to-emulate-gps-location-in-the-android-emulator
While on real devices the location is emulated via https://stackoverflow.com/questions/40414011/how-to-set-the-android-6-0-mock-location-app-from-adb

**Reply by itkhanz** [reply]
> Both emulators and real devices use different approach to emulate geolocation. Emulators utilize https://stackoverflow.com/questions/2279647/how-to-emulate-gps-location-in-the-android-emulator While on real devices the location is emulated via https://stackoverflow.com/questions/40414011/how-to-set-the-android-6-0-mock-location-app-from-adb

Thank you. I searched deeper and found this explained in README of appium settings repo on how Appium utilizes Android internals to mock location:

https://github.com/appium/io.appium.settings?tab=readme-ov-file#setting-mock-locations

**Comment by itkhanz** [comment]
@mykola-mokhnach Actually my Android app interacts with  location permissions `Allow location while using app` popup and has Google Maps on home screen integrated within app. So my initial thoughts for Maps not loading intermittently was that perhaps it has something to do with mock location in place as I am using Appium [mock location from BrowserStack](https://www.browserstack.com/docs/app-automate/appium/test-real-user-conditions/simulate-gps-location). I also tried to use mock location directly from Appium but I still have this issue where maps do not load sometimes. Not sure at which layer (BrowserStack, Appium mock location, application under test) the problem lies and how to debug it further.
Any suggestions from your side would be helpful
